### ISSUE TITLE
Allow option of "posting" to /new endpoint with a pre-filled jsmd document

### ISSUE THREAD

**Issue by wlach** [iodide-project/iodide#1397]
Right now we have a `/new` endpoint which will auto-create a notebook based on a template inside the iodide repository:

https://github.com/iodide-project/iodide/blob/e2ee77fdde96210eddb29ea756964aeb0af9afcd/server/notebooks/views.py#L103
https://github.com/iodide-project/iodide/blob/e2ee77fdde96210eddb29ea756964aeb0af9afcd/server/notebooks/templates/new_notebook_content.jsmd

What we want to do here is optionally allow posting to the above endpoint with a json payload containing something like `{ 'jsmd': <my cool iodide notebook> }`. If that's the case and the user is logged in, autocreate the notebook using that payload instead of the default. If the user is *not* logged in, we should redirect to the `/tryit/` endpoint with the same payload (which should do exactly the same thing, but not actually create the notebook) -- this depends on #1330 being implemented (hoping to get that done today).

So the initial use case for this is to facilitate exploration of redash queries in iodide (mozilla/redash#426), but there are probably other cases where this might be useful. 

/cc @jezdez @openjck

**Comment by bcolloran** [comment]
Thanks for filing @wlach. @openjck so excited to have you joining in! :-) :100:

**Comment by openjck** [comment]
Question about the JSON payload. Is there an existing format that it should adhere to or is defining that format part of this work?

**Comment by wlach** [comment]
> Question about the JSON payload. Is there an existing format that it should adhere to or is defining that format part of this work?

No existing format. To start with, something simple like this should work:

```json
{
    "name": "My cool notebook",
    "jsmd": "%% md\nFoo bar\n..."
}
```

**Comment by openjck** [comment]
Okay, great!

Another question as I think about this: does Iodide support private notebooks? Given that the data on Redash is private, do we want to implement that feature as part of this work, or we were thinking about addressing this some other way, like setting up a different instance of Iodide behind LDAP?

**Comment by wlach** [comment]
> Another question as I think about this: does Iodide support private notebooks? Given that the data 
> on Redash is private, do we want to implement that feature as part of this work, or we were thinking > about addressing this some other way, like setting up a different instance of Iodide behind LDAP?

Well, I wouldn't say the data on redash is private *by definition*, we sometimes share portions of it with the public using the API (chutten's crash dashboard would be a previous example of this, there might be others). 

That said, we are setting up an internal iodide server for exactly this use case. Currently just a staging site, but we'll probably promote a version to production soon:

https://data-iodide.stage.mozaws.net/ 

It would be nice if we somehow had the option of creating a notebook on either, depending on the confidentiality requirements. It would be nice to encourage people to be open/public with their analyses/reports where possible.

**Comment by openjck** [comment]
I had another chat with William about this yesterday. I was confused. For now at least, we don't want to accept data over POST, only JSMD. The */new* endpoint should accept JSMD over post and create a notebook with that JSMD. Redash will send JSMD which includes code for loading Redash data.

**Comment by jezdez** [comment]
@wlach @openjck Ah, I didn't realize you were taking a stab at this as well since Rob asked me to look at this next. Any chance we can have a meeting together some time to sync notes?

**Comment by openjck** [comment]
Sure! I put an invite on both of your calendars for 10am Eastern / 4pm Berlin on Thursday. Let me know if that doesn't work.

**Comment by openjck** [comment]
Scratch that. I moved it to 9:30am Eastern / 15:30 Berlin. If that's too short of notice, Will, just let me know and we can reschedule for another time Thursday or some time on Friday.

**Comment by openjck** [comment]
We discussed another important component of the user experience for this feature in the meeting today: we should be sure that the notebook that Redash creates has thorough documentation. There are two types of documentation that we want to provide:

* Documentation for the data itself (title, description if any, how to get back to the raw data on Redash, etc.)
* Documentation for the code (how it's fetching data from Redash, etc.)

**Comment by wlach** [comment]
Talking to @openjck on gitter, we realized that current session-based authentication to protect/use the endpoint isn't going to work, since neither redash nor the web connection to it will actually have the credentials to post to our server (the former just doesn't have them, the latter would be restricted from contacting iodide due to cors).

To make this work, I think we'll need to allow a type of API client which can communicate with the iodide server via API key/token rather than a user session. For [treeherder](https://github.com//mozilla/treeherder) we used hawk/[hawkrest](https://hawkrest.readthedocs.io/) and it seemed to work pretty well. We had a seperate table of authorized submitters + their credentials who had permission to post job information to our servers. This was later deprecated in favor of another solution (another story) but I think this will work pretty well in a federated environment like the one we're proposing for redash or iodide. 

Conveniently, @edmorley just landed a PR taking this functionality out, which you can reverse to see what would need to be added to iodide:

https://github.com/mozilla/treeherder/pull/4075/files

In the context of this bug, I think we would modify the post request to optionally accept a user argument when the request is made using hawk, which would let the query be made *on behalf* of a user on the redash server. The redash server would then return a url to the created query, which the client could then open in a new tab.

Another good reference would be django rest framework's authentication section: https://www.django-rest-framework.org/api-guide/authentication/

@jezdez curious if you have any thoughts on this
@openjck assuming no one object's too strenuously, maybe we can go through what would be involved in implementing this early next week

**Comment by jezdez** [comment]
I'm at a workshop right now and can go into details (tomorrow) but my perspective is:

1. we shouldn't push the data directly from the Redash frontend (requiring a big overhead with handling authentication and authorization) but a Celery task running in Redash that simply can use a shared secret (e.g. drf's built-in [TokenAuthentication](https://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication))
2. the Redash frontend extension itself will be covered by Redash's own authentication system when requesting the sending to the iodide server from the frontend from an own request handler (advantage being that the celery task can do heavy lifting of preparing the data to send) 
3. ask sec to make a statement about which auth scheme is considered state of the art since I have the gut feel that Hawk isn't anymore. I don't want to decide on something that we regret just to get a prototype running

**Comment by edmorley** [comment]
hawkrest is not actively maintained, so I suspect not the best choice any more unfortunately.
I would use whatever is most maintained for Django + considered ok by the security team.

**Comment by wlach** [comment]
Oops, yes, sounds like drf's native TokenAuthentication is the way to go, not hawk. I think some of the treeherder work may still be relevant. Thanks @edmorley and @jezdez.

@jezdez I feel like using celery here is probably overkill -- can't we just do the post as part the request? It should only take a few seconds at most.

**Comment by jezdez** [comment]
@wlach Hm, sure at first we can do the post right from the new Redash API handler that we'll have to write anyway. But Redash outsources other heavy lifting code to Celery tasks for a reason, to reduce the chance of blocking the Python process (given the amount of requests at the REST API it receives). So in case the payloading building and submission to Iodide takes long (for whatever reason) we should consider moving the code to a Celery tasks later on -- totally something that we can file under the "deploy to production" goal in Q2.

**Comment by openjck** [comment]
I'm going to close #1499 since it needs a fair amount of additional work, but the discussion in there is super helpful.

**Comment by wlach** [comment]
After investigating further, I don't think this is actually what we want. Instead, we should just allow suitably-authorized clients to create the notebook using the standard iodide APIs. They should get enough metadata in the response to be able to direct the user to any created notebook:

See #1642
